cdap-app-templates/cdap-etl/cdap-calcite/src/main/java/io/cdap/cdap/datapipeline/relational/CalciteSQLDialectConverter.java

```diff
@@ -0,0 +1,116 @@
+import org.apache.calcite.schema.SchemaPlus;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.dialect.BigQuerySqlDialect;
+import org.apache.calcite.sql.dialect.PostgresqlSqlDialect;
+import org.apache.calcite.sql.dialect.SparkSqlDialect;
+import org.apache.calcite.sql.parser.SqlParseException;
+import org.apache.calcite.sql.parser.SqlParser;
+import org.apache.calcite.sql.validate.SqlConformance;
+import org.apache.calcite.sql.validate.SqlConformanceEnum;
+import org.apache.calcite.sql.validate.SqlValidator;
+import org.apache.calcite.tools.FrameworkConfig;
+import org.apache.calcite.tools.Frameworks;
+import org.apache.calcite.tools.ValidationException;
+
+import javax.annotation.Nullable;
+
+public class CalciteSQLDialectConverter implements SQLDialectConverter {
+    @Override
+    public void validate(String expression, SQLDialect srcDialect, @Nullable Schema schema) throws SQLDialectException {
+        PlannerImpl planner = getPlanner(srcDialect, schema);
+
+        try {
+            planner.validate(planner.parse(expression));
+        } catch (SqlParseException | ValidationException exception) {
+            throw new SQLDialectException(exception.getMessage(), exception);
+        }
+    }
+
+    @Override
+    public String convert(String expression, SQLDialect srcDialect, SQLDialect destDialect, @Nullable Schema schema)
+            throws SQLDialectException {
+        PlannerImpl planner = getPlanner(srcDialect, schema);
+
+        SqlNode node;
+        try {
+            node = planner.parse(expression);
+        } catch (SqlParseException exception) {
+            throw new SQLDialectException(exception.getMessage(), exception);
+        }
+        return node.toSqlString(getDialectObject(destDialect)).getSql();
+    }
+
+    private PlannerImpl getPlanner(SQLDialect srcDialect, @Nullable Schema schema) throws SQLDialectException {
+        if (schema == null) {
+            throw new SQLDialectException("Calcite-based Sql parser / validator requires an input schema, none found");
+        }
+
+        CDAPCalciteSchema calciteSchema = CDAPCalciteSchema.fromCDAPSchema(schema);
+        SchemaPlus schemaPlus = Frameworks.createRootSchema(true);
+        schemaPlus = schemaPlus.add(schema.getRecordName(), calciteSchema);
+
+        SqlParser.Config parserConfig = SqlParser.config()
+                .withCaseSensitive(false)
+                .withConformance(getConformanceFromDialect(srcDialect));
+
+        SqlValidator.Config validatorConfig = SqlValidator.Config.DEFAULT
+                .withConformance(getConformanceFromDialect(srcDialect));
+
+        final FrameworkConfig config = Frameworks.newConfigBuilder()
+                .defaultSchema(schemaPlus)
+                .parserConfig(parserConfig)
+                .sqlValidatorConfig(validatorConfig)
+                .build();
+        PlannerImpl planner = new PlannerImpl(config);
+
+        return planner;
```

Comment:
Should this result be cached? Is it possible for `validate` to be called multiple times over the same schema?

Reply:
`validate()` will only be called once per pipeline stage I believe, so I don't think it is terribly necessary to cache the planner.